discovery-provider/alembic/versions/e68928955812_nfc_unicode_routes.py

```diff
@@ -0,0 +1,149 @@
+"""nfc_unicode_routes
+
+Revision ID: e68928955812
+Revises: 9b4e12c0c428
+Create Date: 2023-02-23 02:09:31.901048
+
+"""
+import unicodedata
+
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.dialects.postgresql import ARRAY
+from sqlalchemy.types import Integer, String
+
+# revision identifiers, used by Alembic.
+revision = 'e68928955812'
+down_revision = '9b4e12c0c428'
+branch_labels = None
+depends_on = None
+
+
+def track_route_updates(connection):
+    """
+    Get all track routes. Check if the NFC normalized slug matches the value in the database.
+    If no match, rewrite the NFC value to the database.
+    """
+    track_routes = connection.execute(
+        """
+            SELECT
+                track_id,
+                slug,
+                title_slug
+            FROM track_routes
+            WHERE
+                is_current = True
+        """
+    )
+    track_ids = []
+    slugs = []
+    title_slugs = []
+    for route in list(track_routes):
+        [track_id, slug, title_slug] = route
+        nfc_slug = unicodedata.normalize('NFC', slug)
+        if nfc_slug != slug:
+            nfc_title_slug = unicodedata.normalize('NFC', title_slug)
+            track_ids.append(track_id)
+            slugs.append(nfc_slug)
+            title_slugs.append(nfc_title_slug)
+
+    if len(track_ids) == 0: # no updates
+        return
+
+    params = {
+        "track_ids": track_ids,
+        "slugs": slugs,
+        "title_slugs": title_slugs
+    }
+
+    sql = sa.text(
+        """
+            UPDATE track_routes SET
```

Comment:
doesn't that not solve our problem then? discovery nodes won't show the right api results if is current rows aren't updated. 

am i misunderstanding?

Reply:
Routes don't need to be current to resolve properly - just having a route for the normalized version will allow the API to resolve the track, is_current or not

Reply:
You're right that the permalink won't change, but both the non-normalized and normalized urls will resolve, so sending the original url via a normalizing client like Slack or Twitter will now work, and sending via Discord and others will continue to work

Reply:
ah I see.

so the best solution then would be to change the current is_current rows to false and then add new is_current rows for all of them. that updates the permalinks, backlinks still work, etc. 

however, I'm not sure what values make sense to add for blockno, txhash, etc with those making any sense... it seems not the worst to break 3K already-half-broken links to fix it going forward for them and make the db consistent with what it will look like going fwd. 

wdyt? interesting Q.

Reply:
Yea exactly RE: best case solution

I think we can dupe whatever tx/block number the existing current route has. Shouldn't be a problem afaik wrt keys/indexes, and I think makes sense.

I don't think we _need_ permalinks to be the normalized one, as long as non-normalized routes are retained, so I don't think we bother updating the `is_current`

Reply:
<img width="966" alt="Screenshot 2023-02-28 at 6 23 27 PM" src="https://user-images.githubusercontent.com/2731362/222028917-f619ccc4-7776-4847-9ba5-846201b344dd.png">


verbose, but it works. i like the verbosity.

Reply:
I guess this is wrong, collision id should be 0 since these are different slugs. Will make that update.

Reply:
spoke offline -- collision id should stay the same so that future edits to the track add onto the existing id